src/lib/premium-computations.js

```diff
@@ -0,0 +1,149 @@
+const { BigNumber, ethers } = require('ethers');
+const { bnMax } = require('./helpers');
+
+const { MaxUint256, WeiPerEther, Zero } = ethers.constants;
+
+const MIN_UNIT_SIZE = WeiPerEther;
+const UNIT_DIVISOR = 1000;
+
+const {
+  PRICE_CHANGE_PER_DAY,
+  SURGE_PRICE_RATIO,
+  SURGE_THRESHOLD_DENOMINATOR,
+  SURGE_THRESHOLD_RATIO,
+  TARGET_PRICE_DENOMINATOR,
+} = require('./constants');
+const calculateBasePrice = (targetPrice, bumpedPrice, bumpedPriceUpdateTime, now) => {
+  const elapsed = now.sub(bumpedPriceUpdateTime);
+  const priceDrop = elapsed.mul(PRICE_CHANGE_PER_DAY).div(3600 * 24);
+  return bnMax(targetPrice, bumpedPrice.sub(priceDrop));
+};
+
+const calculateFixedPricePremiumPerYear = (coverAmount, price) => {
+  return coverAmount.mul(price).div(TARGET_PRICE_DENOMINATOR);
+};
+
+const calculatePremiumPerYear = (coverAmount, basePrice, initialCapacityUsed, totalCapacity) => {
+  const basePremium = coverAmount.mul(basePrice).div(TARGET_PRICE_DENOMINATOR);
+  const finalCapacityUsed = initialCapacityUsed.add(coverAmount);
+  const surgeStartPoint = totalCapacity.mul(SURGE_THRESHOLD_RATIO).div(SURGE_THRESHOLD_DENOMINATOR);
+
+  if (finalCapacityUsed.lte(surgeStartPoint)) {
+    return basePremium;
+  }
+
+  const amountOnSurgeSkip = initialCapacityUsed.sub(surgeStartPoint).gt(0)
+    ? initialCapacityUsed.sub(surgeStartPoint)
+    : Zero;
+
+  const amountOnSurge = finalCapacityUsed.sub(surgeStartPoint);
+  const totalSurgePremium = amountOnSurge.mul(amountOnSurge).mul(SURGE_PRICE_RATIO).div(totalCapacity).div(2);
+  const skipSurgePremium = amountOnSurgeSkip.mul(amountOnSurgeSkip).mul(SURGE_PRICE_RATIO).div(totalCapacity).div(2);
+  const surgePremium = totalSurgePremium.sub(skipSurgePremium);
+
+  return basePremium.add(surgePremium);
+};
+
+/**
+ * This function allocates each unit to the cheapest opportunity available for that unit
+ * at that time given the allocations at the previous points.
+ *
+ * To solve the allocation problem we split the amount A in U = `A / UNIT_SIZE + 1`
+ * units and use that for the rest of the computations.
+ *
+ * To allocate the units U we take each unit at a time from 0 to U,
+ * and allocate each unit to the *best* pool available in terms of price
+ * as long as the pool has enough capacity for that unit.
+ *
+ * If there is no pool with capacity available for a unit of allocation the function returns
+ * with an empty list (aborts allocation completely - no partial allocations supported).
+ *
+ * UNIT_SIZE is dynamically computed as 0.1% (1 / UNIT_DIVISOR) of the cover amount.
+ * It has a minimum which is 1e18 (can't go below 1 ETH or 1 DAI).
```

Comment:
0.1% seems way too small, last we discussed we should split the cover in 10. Even if we split it in 10 parts - we should have a minimum, currently we're eyeballing it to be somewhere between 10k and 100k DAI (or equivalent). Needs a bit of research by going through covers and comparing gas costs and premiums since the gas cost should not make the cover overly expensive.

Reply:
what's the downside to using say 1% instead of 10% if the computation still completes fast enough?

Reply:
Downside of using units that are too small is that we may end up allocating too many pools and the gas costs would offset the optimizations that we obtain from "efficiently" routing.

Reply:
1 DAI is too low as a chunk size imo - imagine having a segment with 1 DAI in it. calibrating the right min amount depends on the gas costs, as Anatol suggested above. my gut feeling is that something like 5K - 10K DAI is more appropriate

Reply:
adressed using 10%